File: cpp_opencv/transformation/transformation.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <stdio.h>
#include <string>

using namespace cv;

Mat translation (Mat img, float x, float y);
Mat rotation (Mat img, int angle, Point2f center);
std::string photo = "../../photos/dog2.jpg";

int main(){
    Mat image = imread(photo);
    imshow("dog", image);


    // Translated 
    Mat translated = translation( image, 100, 100 );
    imshow("translated", translated);

    // Rotation
    Mat rotated = rotation ( image, 45, Point2f(image.cols / 2, image.rows /2 ));
    imshow("rotated", rotated);

    // Flipping 
    Mat flipped;
    flip(image, flipped, 0);
    imshow("flipped", flipped);

    waitKey(0);
    return 0;
}

Mat translation(Mat img, float x, float y){
    Mat rtn;
    float warpvalues [] = { 1.0, 0.0, x, 0.0, 1.0, y };
    Mat translationMatrix = Mat(2, 3, CV_32F, warpvalues);

    warpAffine(img, rtn, translationMatrix, img.size());
    return rtn;
}

Mat rotation (Mat img, int angle, Point2f center = Point2f(0, 0)){
    Mat rtn;
    Mat rotationMatrix = getRotationMatrix2D(center, angle, 1);

    warpAffine(img, rtn, rotationMatrix, img.size());
    return rtn;
}   

File: py_opencv/transformation.py
import cv2 as cv
import numpy as np

img = cv.imread('photos/dog2.jpg')

cv.imshow('dog', img)

"""Translation"""
def translate(img, x, y):
    transMat = np.float32([[1, 0, x], [0,1,y]])
    dimensions = (img.shape[1], img.shape[0]) # width, height
    return cv.warpAffine(img, transMat, dimensions)

"""Rotation"""
def rotate(img, angle, roPoint = None):
    (height, width) = img.shape[:2]
    if roPoint == None:
        roPoint = (width//2, height//2)
    rotMat = cv.getRotationMatrix2D(roPoint, angle, 1.0)
    dimenstions = (width, height)
    return cv.warpAffine(img, rotMat, dimenstions)

# TRANSLATION neg x = left, neg y = up
translatedImg = translate(img, -100, 100)
cv.imshow('translated', translatedImg)

# ROTATION neg = clockwise
rotated = rotate(img, -45)
cv.imshow('rotated', rotated)
rotated_rotated = rotate(rotated, 45)
cv.imshow('rotated-rotated', rotated_rotated)

# RESIZING
resized = cv.resize(img, (500, 500), interpolation=cv.INTER_LINEAR)
cv.imshow('resized', resized)

# FLIPPING
flipped = cv.flip(img, -1) # flip code 0 = vertical flip, 1 = horizontal flip, -1 = both
cv.imshow('flipped', flipped)

# CROPPING
cropped = img[200:400, 300:400]
cv.imshow('cropped', cropped)

cv.waitKey(0)